Centralize storage and updating of Sonos favorites

jjlawren opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Favorites in Sonos are unique per "household" and do not need to be updated and stored separately for every speaker. They're also tied to update identifiers which we can use to conditionally update our cache when changes actually occur. This should help avoid possible race conditions and eliminate the need for bandaids like in #50495 and #50575.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

jjlawren committed:
Centralize storage and updating of Sonos favorites

jjlawren commented:
I'm finding we sometimes request the updated favorites too quickly when receiving an event. Perhaps there needs to be a slight delay before requesting the current known favorites.

bdraco commented on homeassistant/components/sonos/__init__.py:
```diff
@@ -122,10 +124,12 @@ def _discovered_player(soco: SoCo) -> None:
+                    _LOGGER.debug("Adding new speaker: %s", speaker_info)
                     speaker = SonosSpeaker(hass, soco, speaker_info)
                     data.discovered[soco.uid] = speaker
+                    data.favorites.setdefault(soco.household_id, SonosFavorites(soco))
```
Can the household_id be different?  I guess you could have two systems on the same network??

jjlawren replied:
Yes, it's rare but can happen. For example, if you have separate S1 and S2 systems on the same network.

jjlawren committed:
Add a delay when updating favorites from an event

bdraco commented on homeassistant/components/sonos/favorites.py:
```diff
@@ -0,0 +1,73 @@
+            return
+
+        self._event_version = event_id
+        self.hass.helpers.event.async_call_later(5, self.update)
```
If it changes 2x within 5s you'll end up with multiple async_call_later's scheduled

jjlawren committed:
Reschedule update if pending, reduce timer to 3s

jjlawren commented:
Although this method is much more efficient, it's also less resilient as it only stores one `SoCo` instance. I have a plan to address this in a followup PR.

bdraco commented on homeassistant/components/sonos/favorites.py:
```diff
@@ -0,0 +1,79 @@
+
+    def __iter__(self) -> Iterator:
+        """Return an iterator for the known favorites."""
+        return iter(self._favorites)
```
Is there any risk of this changing during iteration?

jjlawren replied:
It would be extremely rare, but I suppose there is a possibility of someone hitting that eventually.

bdraco replied:
Can we return a copy of the list and iterate that instead ?

jjlawren replied:
Probably safer to return a copy. Will adjust.

jjlawren committed:
Copy list before iterating

jjlawren committed:
Add more typing

jjlawren committed:
Convert to callback

jjlawren committed:
Add resilience to favorites updating

Convert discovered to an OrderedDict
Fallback to other speakers if updating favorites fails
Prefer any working speaker during next update
Don't store own SoCo instances in SonosFavorites

jjlawren commented:
> Although this method is much more efficient, it's also less resilient as it only stores one `SoCo` instance. I have a plan to address this in a followup PR.

Made sense to just handle it in this PR. Should be ready for review now.

MartinHjelmare commented on homeassistant/components/sonos/favorites.py:
```diff
@@ -0,0 +1,95 @@
+
+    def __init__(self, hass: HomeAssistant, household_id: str) -> None:
+        """Initialize the data."""
+        self.hass: HomeAssistant = hass
```
```suggestion
        self.hass = hass
```

jjlawren replied:
Good to know 👍

MartinHjelmare commented on homeassistant/components/sonos/favorites.py:
```diff
@@ -0,0 +1,95 @@
+    def __init__(self, hass: HomeAssistant, household_id: str) -> None:
+        """Initialize the data."""
+        self.hass: HomeAssistant = hass
+        self.household_id: str = household_id
```
```suggestion
        self.household_id = household_id
```

MartinHjelmare commented on homeassistant/components/sonos/speaker.py:
```diff
@@ -128,6 +129,7 @@ def __init__(
         """Initialize a SonosSpeaker."""
         self.hass: HomeAssistant = hass
         self.soco: SoCo = soco
```
```suggestion
        self.hass = hass
        self.soco = soco
```

MartinHjelmare reviewed (commented):
If we type the parameter in the signature we don't need to type the same variable again in the function body.

jjlawren committed:
Apply typing suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

frenck reviewed (approved):
LGTM ✅ 

Tested with multiple Sonos speakers, seems to be working fine.

../Frenck 🚂

jjlawren committed:
Add type for new attribute, comment for non-obvious type choice
